src/goto-symex/symex_complexity.cpp

```diff
@@ -0,0 +1,197 @@
+/*******************************************************************\
+
+Module: Symbolic Execution
+
+Author: John Dumbell
+
+\*******************************************************************/
+
+#include "symex_complexity.h"
+#include "goto_symex_state.h"
+
+/// Takes into account nested loops and various other sublties.
+optionalt<loop_complexity_infot>
+symex_complexityt::get_current_active_loop(call_stackt &current_call_stack)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    // We want to find the most deeply-nested active loop in a frame.
+    optionalt<loop_complexity_infot> active_loop;
+    for(auto &loop_iter : frame_iter->loop_iterations)
+    {
+      if(loop_iter.second.is_active)
+      {
+        if(
+          !active_loop || active_loop->loop_info->nesting_level.force() <
+                            loop_iter.second.nesting_level.force())
+        {
+          active_loop =
+            loop_complexity_infot(loop_iter.first, &loop_iter.second);
+        }
+      }
+    }
+
+    // As we're walking bottom-up, the first frame we find with active loops
+    // is our closest active one.
+    if(active_loop)
+      return active_loop;
+  }
+
+  return {};
+}
+
+bool symex_complexityt::in_blacklisted_loop(
+  call_stackt &current_call_stack,
+  const irep_idt &loop_id)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    for(auto &loop_iter : frame_iter->loop_iterations)
+    {
+      if(
+        loop_iter.second.is_active &&
+        loop_iter.second.blacklisted_loops.count(loop_id) != 0)
+      {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool symex_complexityt::are_loop_children_too_complicated(
+  call_stackt &current_call_stack)
+{
+  // Check that current loops don't violate complexity boundaries, if they
+  // do then add that loop to the blacklist for all currently-running loops.
+
+  // We want to make sure that if a certain execution of a loop is too complex
+  // no matter the branching, that we cancel out early. This is critical because
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+  irep_idt blacklist;
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    std::vector<loop_complexity_infot> info_vectors;
+    for(auto &loop_iter : frame_iter->loop_iterations)
+    {
+      if(loop_iter.second.is_active)
+        info_vectors.emplace_back(loop_iter.first, &loop_iter.second);
```

Comment:
:pick: info_vectors could be made `const` if using `make_range(frame_iter->loop_iterations).filter(...).map(...)`